EuiDataGrid add column sorting

chandlerprall opened the pull request:
### Summary

Adds a sorting prop to describe the currently state of sort and a callback function when the state is changed by user interaction. By default the grid does not sort anything itself, but this can be opted into by passing the `inMemory` prop, demonstrated by a new example in the docs.

![in memory sorting](https://d.pr/i/U3oWPt+)

myasonik commented on src/components/datagrid/data_grid_header_row.tsx:
```diff
@@ -33,9 +39,30 @@ const EuiDataGridHeaderRow: FunctionComponent<
+        } = {};
+        if (
+          sorting &&
+          sorting.columns.length === 1 &&
```
If the `length > 2` can we add an `aria-describedby` attribute on the cell?

None of this is defined in the spec so we just need to make up a good experience. This seems like a good starting place to me though I'm open to any other suggestions too.

Pseudo-code simplified example:
```
const multiSort = sorting.columns.length > 1;
const ariaDescribeBy =  multiSort ? { 'aria-describedby': randomId() } : {};
const sortString = sorting.columns.map(col => `Sort by ${col.name} ${col.direction}`).join(' then ');
}, '');

return (
  <div role="columnheader" {...ariaDescribeBy}>
    Column name
	{multiSort && 
      <EuiScreenReaderOnly id={randomId}>
		{sortString}
      </EuiScreenReaderOnly>
    }
  </div>
)
```

myasonik commented:
The in memory table never gets the top left cell `tabIndex` set to `0` by default so it's impossible to tab into (without first interacting with it with a mouse). 

I tested it with just the memory table rendered, not only 2 tables on one page, so I don't think it's the same focus issues that @snide is hitting in #2230.

chandlerprall commented:
> The initial state has `[{ id: 'contributions', direction: 'asc' }]` and the toggles correctly show that 'contributions' should be in ascending order, but the column looks to be unsorted.

@thompsongl currently the in-memory sort is purely text based and ends up as a lexicographical sort. As an example:

```javascript
['2', '10', '3', '13'].sort() // ["10", "13", "2", "3"]
```

With the upcoming schema changes the sorting will know what kind of data is being sorted and will be smarter.

thompsongl commented on src/components/datagrid/data_grid_inmemory_renderer.tsx:
```diff
@@ -0,0 +1,69 @@
+
+  return createPortal(
+    <Fragment>{rows}</Fragment>,
+    (documentFragment as unknown) as Element
```
Why a document fragment?

chandlerprall replied:
Could have used a `div` instead, but a document fragment feels more correct to me? The main goal is to not include the resulting DOM in the main window/document.

chandlerprall commented:
Fixed up the sorting after @snide's merge and addressed the tab-into and arrow key navigation @myasonik found.

There's a performance problem caused by having the focus trap on every cell, but I'm leaving that for another PR.

This is ready for another review pass.

chandlerprall commented:
@thompsongl good catch! I'm going to come back around to that in another PR, need to clean up how the focused cell is tracked.

chandlerprall commented:
As @myasonik is out and no other (sorting-related) issues have presented, I'm merging this into the feature branch to allow other cleanups to begin.
